### DISCUSSION TITLE
TLS/SSL Authentication

### DISCUSSION THREAD

**Discussion by smhmaqsood** [arendst/Tasmota#11834] [Q&A] [answered]
Hi,

I've been working on tasmota and I need help regarding TLS/SSL on ESP8266 1MB(FS:none; OTA:502kb).
I want to configure it but I'm totally blank about it.
I've gone through documents and discussions but these things are not clear.
https://tasmota.github.io/docs/TLS/
https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/
http://www.steves-internet-guide.com/mosquitto-tls/

I Want To Know How Can I Configure TLS on ESP8266 from scratch...?
I would really appreciate your help.

**Comment by chaveiro** [comment, marked as answer]
1 - Compile with this set on user_config_override.h
#ifndef USE_MQTT_TLS 
#define USE_MQTT_TLS                             // Use TLS for MQTT connection (+34.5k code, +7.0k mem and +4.8k additional during connection handshake)
 #undef MQTT_TLS_ENABLED
 #define MQTT_TLS_ENABLED       true             // [SetOption103] Enable TLS mode (requires TLS version)

2- Have your mqtt server running and supporting TLS and with a certificate configured. (I use mosquitto mqtt),

3 - Setup your server settings and activate TLS from the tasmota web portal.

4 - Then issue this from the console to auto learn the fingerprint of your server:
MqttFingerprint1 00

Should now connect to mqtt with TLS.

**Comment by hnaeem07** [comment]
Hello Guys, I want built my TLS enabled tasmota bin file without webserver (//#define USE_WEBSERVER) , i cannot enter the TLSKeys1 and TLSKeyss2 on the serial interface, plus mqtt is not connected since its TLS enabled. can hardcode the Key files ?

Really appreciate if anyone could help. 

Thank you in advance.

**Reply by stefanbode** [reply]
There is/was a dummy key that allow a connection on the fist time and then store the key to flash. Instill believe this is the default. Therefore on 8883 you should be able to connect. There is a userconfig_override. Here you can set nearly all parameters. E.g. wlan credentials if you like.

**Reply by stefanbode** [reply]
https://tasmota.github.io/docs/TLS/

Here you should find all answers you need to get it working.

**Comment by hnaeem07** [comment]
Thank you so much for the reply, this is helpful and will try, dummy key is in the default Tasmota firmware or I have to enter keys from letsencrypt / easyrsa to the device?

**Reply by hnaeem07** [reply]
Thanks a lot @stefanbode.  tried as in the document, and works, with webserver enabled, with fingerprint option,  entering TLSKeys 1 and 2, but cannot get get this connected with the broker without webserver interface.

**Reply by stefanbode** [reply]
As described you do not enter something when using the fingerprint. All the 0x00 should be fine or can connect to any mqtt with TLS. After the first connect the fingerprint is stored for future connects.

**Comment by hnaeem07** [comment]
Great, thanks a lot for the clarification,  appreciated. 👍
